src/ray/raylet/worker_pool.cc

```diff
@@ -955,6 +958,8 @@ void WorkerPool::TryKillingIdleWorkers() {
         rpc::ExitRequest request;
         rpc_client->Exit(request, [this, worker](const ray::Status &status,
                                                  const rpc::ExitReply &r) {
+          RAY_CHECK(pending_exit_idle_workers_.count(worker->WorkerId()));
```

Comment:
Why do we move it up here?

Reply:
Sounds good. By the way, you don't need the `count` check in addition to the erase check.

Reply:
Also, can you add a test for the no leaks?

You can do this by adding an AssertNoLeaks() method to the end of all the C++ tests. We've been doing this incrementally for other classes, probably a good time to add this to the WorkerPool too.

Reply:
@stephanie-wang I'm not sure if `AssertNoLeaks` is doable in the worker pool test. Sounds to me `AssertNoLeaks` requires a graceful cleanup in every test case, but it's not true in the worker pool test. e.g. we don't unregister workers at the end of each test case.

Reply:
To be clear, this is not fixing the existing issue, but it is here to make this function more robust right? Afaik, moving this check up here will have the same behavior as before because the function doesn't return in the middle.

Reply:
OK, let's do the AssertNoLeaks just for the new test cases, though.

It should always be possible to check for graceful cleanup. If we can't do it, then it's a sign that we need to improve the software design.

Reply:
@rkooo567 I didn't realize that you guys removed the return statement in `if (!status.ok())`. Our internal codebase is several versions behind 🤣 . In this case, indeed moving it to the very beginning of the callback only prevents potential bugs in future development.

Reply:
@stephanie-wang I think I can manage to add `AssertNoLeaks` for this case, but I have one question. In the current worker pool design, there are idle workers in most cases. Currently, I don't think that we have a way to clean up all idle and non-idle workers gracefully in non-testing scenarios. I believe that even if the driver of a job has exited and all tasks and actors have finished or been destroyed, there are still idle workers for this job running, unless `TryKillingIdleWorkers` kills them partially or completely, based on the number of non-idle workers on this node. So, what's your idea to clean them at the end of the test case? (Please correct me if I misunderstand anything.)

Reply:
Hmm I see, how about just checking that the pending_exit_idle_workers_ map gets cleaned up?